src/bgw/job_stat.c

```diff
@@ -0,0 +1,348 @@
+} JobResult;
+
+
+static
+TimestampTz
+calculate_next_start_success(TimestampTz last_finish, BgwJob *job)
+{
+	/* TODO: add randomness here? Do we need a range or just a percent? */
+	TimestampTz ts = DirectFunctionCall2(timestamptz_pl_interval,
+										 TimestampTzGetDatum(last_finish),
+										 IntervalPGetDatum(&job->fd.schedule_interval));
+
+	return ts;
+}
+
+/* For failures we have standard exponential backoff based on consecutive failures
+ * along with a ceiling at schedule_interval * MAX_INTERVALS_BACKOFF */
+static
+TimestampTz
+calculate_next_start_failure(TimestampTz last_finish, int consecutive_failures, BgwJob *job)
+{
+	/* TODO: add randomness here? Do we need a range or just a percent? */
+	/* consecutive failures includes this failure */
+	float8		multiplier = 1 << (consecutive_failures - 1);
+
+	/* ival = retry_period * 2^(consecutive_failures - 1)  */
+	Datum		ival = DirectFunctionCall2(interval_mul, IntervalPGetDatum(&job->fd.retry_period), Float8GetDatum(multiplier));
+
+	/* ival_max is the ceiling = MAX_INTERVALS_BACKOFF * schedule_interval */
+	Datum		ival_max = DirectFunctionCall2(interval_mul, IntervalPGetDatum(&job->fd.schedule_interval), Float8GetDatum(MAX_INTERVALS_BACKOFF));
+
+	if (DatumGetInt32(DirectFunctionCall2(interval_cmp, ival, ival_max)) > 0)
+		ival = ival_max;
+
+	return DatumGetTimestampTz(DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
+}
+
+/* For crashes, the logic is the similar as for failures except we also have
+*  a minimum wait after a crash that we wait, so that if an operator needs to disable the job,
+*  there will be enough time before another crash.
+*/
+static
+TimestampTz
+calculate_next_start_crash(int consecutive_crashes, BgwJob *job)
+{
+	TimestampTz now = GetCurrentTimestamp();
+	TimestampTz failure_calc = calculate_next_start_failure(now, consecutive_crashes, job);
+	TimestampTz min_time = TimestampTzPlusMilliseconds(now, MIN_WAIT_AFTER_CRASH_MS);
+
+	if (min_time > failure_calc)
+		return min_time;
+	return failure_calc;
+}
+static bool
+bgw_job_stat_tuple_mark_end(TupleInfo *ti, void *const data)
+{
+	JobResult  *res = data;
+	HeapTuple	tuple = heap_copytuple(ti->tuple);
+	FormData_bgw_job_stat *fd = (FormData_bgw_job_stat *) GETSTRUCT(tuple);
+	Interval   *duration;
+
+	fd->last_finish = GetCurrentTimestamp();
+
+
+
+	duration = DatumGetIntervalP(DirectFunctionCall2(timestamp_mi, TimestampTzGetDatum(fd->last_finish), TimestampTzGetDatum(fd->last_start)));
+	fd->total_duration = *DatumGetIntervalP(DirectFunctionCall2(interval_pl, IntervalPGetDatum(&fd->total_duration), IntervalPGetDatum(duration)));
+
+
+	/* undo marking created by start marks */
+	fd->last_run_success = res->success;
+	fd->total_crashes--;
+	fd->consecutive_crashes = 0;
+
+	if (res->success)
+	{
+		fd->total_success++;
+		fd->consecutive_failures = 0;
+		/* Mark the next start at the end if the job itself hasn't */
+		if (!bgw_job_stat_next_start_was_set(fd))
```

Comment:
If you put `calculate_next_start_success` and `calculate_next_start_crash` in the same function, then you only have to do this `if` statement once outside of the `if (res->success)` statement.

Reply:
as above